### PR TITLE
enhancement(http source): Add `gzip` & `deflate` decompression

### LINKED ISSUE #4218: Automatic decompression in the `http` source
Given the appropriate [`Content-Encoding` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) the `http` source should automatically decompress the incoming data.

## Requirements

- [ ] `Content-Encoding: gzip`
- [ ] `Content-Encoding: deflate`
- [ ] combined, ex: `Content-Encoding: deflate, gzip`

### PR DESCRIPTION
**Pull request by ktff**
Closes #4218 

Also adds support for `snappy` decompression.

### Todo

- [ ] Follow up on https://github.com/timberio/vector/pull/5638#discussion_r546706940
- [x] Add integration test with at least one remote_write sender.
- [x] Somehow verify that all implementations of remote_write protocol are setting `Content-Encoding`. If this can not be done, then it's better to revert `snappy` decompression back to original source and just pass the data in util http source when `snappy` encoding is detected.

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice. This will need doc updates though.

File: src/sources/util/http.rs [hunk 1]
```diff
@@ -124,6 +125,38 @@ impl HttpSourceAuth {
...
+    if let Some(encodings) = header {
+        for encoding in encodings.rsplit(',').map(str::trim) {
+            body = match encoding {
+                "identity" => body,
+                "gzip" => decode_read(GzDecoder::new(body.reader()), "gzip")?,
+                "deflate" => decode_read(DeflateDecoder::new(body.reader()), "deflate")?,
+                encoding => {
+                    return Err(ErrorMessage::new(
+                        StatusCode::UNSUPPORTED_MEDIA_TYPE,
+                        format!("Unsupported encoding {:?}", encoding),
```
Comment (fanatid) [inline src/sources/util/http.rs:138]: Why debug (`{:?}`) instead of display (`{}`)?

File: src/sources/util/http.rs [hunk 2]
```diff
@@ -156,27 +189,32 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
...
-                        let events = match auth.is_valid(&auth_header) {
-                            Ok(()) => self.build_event(body, headers, query_parameters),
-                            Err(err) => Err(err),
-                        };
+                        let events = auth
+                            .is_valid(&auth_header)
+                            .and_then(|()| decode(&encoding_header, body))
+                            .and_then(|body| {
+                                self.build_event(body.clone(), headers, query_parameters)
+                                    .map(|events| (events, body.len()))
```
Comment (fanatid) [inline src/sources/util/http.rs:212]:
```suggestion
                                let body_size = body.len();
                                self.build_event(body, headers, query_parameters)
                                    .map(|events| (events, body_size))
```
`.len()`  before `build_event` will be cheaper than `.clone()`..

**Review by fanatid — APPROVED** [review]
Code looks good :+1: 
Need green CI and docs update.

File: src/sources/util/http.rs [hunk 3]
```diff
@@ -124,6 +125,38 @@ impl HttpSourceAuth {
     }
 }
 
+fn decode(header: &Option<String>, mut body: Bytes) -> Result<Bytes, ErrorMessage> {
+    if let Some(encodings) = header {
+        for encoding in encodings.rsplit(',').map(str::trim) {
+            body = match encoding {
+                "identity" => body,
+                "gzip" => decode_read(GzDecoder::new(body.reader()), "gzip")?,
```
Comment (jszwedko) [inline src/sources/util/http.rs:133]: Would there be any advantage to using [`async_compression`](https://docs.rs/async-compression/0.3.5/async_compression/) here? We use it in a few other places.

File: src/sources/util/http.rs [hunk 4]
```diff
@@ -124,6 +125,38 @@ impl HttpSourceAuth {
...
+        }
+    }
+
+    Ok(body)
+}
+
+fn decode_read(mut decoder: impl Read, coding: &str) -> Result<Bytes, ErrorMessage> {
+    let mut decoded = Vec::new();
+    decoder.read_to_end(&mut decoded).map_err(|error| {
+        warn!(message = "Failed decoding payload.", %coding, %error);
```
Comment (jszwedko) [inline src/sources/util/http.rs:151]: Should this be an internal event?

File: src/sources/util/http.rs [hunk 5]
```diff
@@ -124,6 +125,38 @@ impl HttpSourceAuth {
     }
 }
 
+fn decode(header: &Option<String>, mut body: Bytes) -> Result<Bytes, ErrorMessage> {
+    if let Some(encodings) = header {
+        for encoding in encodings.rsplit(',').map(str::trim) {
```
Comment (jszwedko) [inline src/sources/util/http.rs:130]: Shouldn't this be `lsplit`?

File: src/sources/util/http.rs (hunk 3 above)
Comment (fanatid) [reply src/sources/util/http.rs:133]: I thought about `async_compression` too but found nothing about using it with warp :thinking: maybe land this and upgrade to `async_compression` as a separate task..?

File: src/sources/util/http.rs (hunk 1 above)
Comment (ktff) [reply src/sources/util/http.rs]: Didn't thought it had `impl Display`.

**Comment by ktff** [comment]
I've moved `snappy` decompression from `prometheus_remote_write` source to shared http util source. Because of it `http` source now also supports `snappy`.

@bruceg move seamed like the best option to resolve the issue of Prometheus setting `Content-Encoding` header as `snappy` so causing us to reject the request since we didn't support it.

File: src/internal_events/http.rs [hunk 6]
```diff
@@ -74,3 +75,24 @@ impl InternalEvent for HTTPEventEncoded {
         counter!("processed_bytes_total", self.byte_size as u64);
     }
 }
+
+#[derive(Debug)]
+pub struct HTTPDecompressError<'a> {
+    pub error: &'a dyn Error,
+    pub coding: &'a str,
```
Comment (jszwedko) [inline src/internal_events/http.rs:82]: I'd normally call this `encoding`

**Review by jszwedko — APPROVED** [review]
This looks good to me!

**Comment by bruceg** [comment]
> @bruceg the move seamed like the best option to resolve the issue of Prometheus setting `Content-Encoding` header as `snappy` so causing us to reject the request since we didn't support that algorithm in shared http source.

If all the remotes send this header properly, then this makes sense. However, the remote_write protocol definition states that remotes are *required* to compress with snappy, and so I am concerned there are going to be some out there that skip the header since it's just implied by the protocol. I'd like to see an integration test with at least one remote_write sender, as we only test this source against ourself.

File: src/sources/util/http.rs [hunk 7]
```diff
@@ -124,6 +126,53 @@ impl HttpSourceAuth {
...
```
Comment (bruceg) [inline src/sources/util/http.rs:131]: I don't see a test for having multiple values here, did I miss it?

File: src/internal_events/http.rs (hunk 6 above)
Comment (ktff) [reply src/internal_events/http.rs]: Yep

File: src/sources/util/http.rs (hunk 7 above)
Comment (ktff) [reply src/sources/util/http.rs:131]: It's called `http_gzip_deflate` in `http` source file.

**Comment by ktff** [comment]
@bruceg a test was added, and we are checking if the proper `Content-Encoding` header is set, and if not, `prometheus_remote_write` source will decode `snappy` it self.

**Pull request merged**